vtex-apps/b2b-organizations-graphql
# Add impersonate user metrics

#### What problem is this solving?

Send events to Analytics Redshift when a user impersonates another user.
Currently, there are two functions used to impersonate and each one will send one different metric (with similar fields) in order to keep tracking about how they are used - there is a comment in the code telling to deprecate one of them, but both are still used)

#### How to test it?
 On [organization manager page](https://takeda--b2bstore005.myvtex.com/account#/organization) select a user to impersonate. After impersonate, click to stop impersonate. Events should appear in Analytics in one hour or so, into "vtex"."schemaless"."b2b_suite_buyerorg_data_raw" schema.



[Workspace](https://takeda--b2bstore005.myvtex.com/account#/organization)

#### Screenshots or example usage:

![image](https://github.com/vtex-apps/b2b-organizations-graphql/assets/1576737/c02266db-f243-41bc-a723-3fdcdd68470e)

![image](https://github.com/vtex-apps/b2b-organizations-graphql/assets/1576737/679ca32e-4f65-4344-b248-fbe42b766da2)


Types:
```
type ImpersonateUserMetric = {
  kind: 'impersonate-user-graphql-event'
  description: 'Impersonate User Action - Graphql'
}

type ImpersonateB2BUserMetric = {
  kind: 'impersonate-b2b-user-graphql-event'
  description: 'Impersonate B2B User Action - Graphql'
}

export type Metric = {
  name: 'b2b-suite-buyerorg-data'
  account: string
} & (ImpersonateUserMetric | ImpersonateB2BUserMetric)


type ImpersonatePerson = {
  email: string
  buy_org_id: string
  cost_center_id: string
  person_id: string
}

type ImpersonateFieldsMetric = {
  user?: ImpersonatePerson
  target: ImpersonatePerson
  date: string
}

type ImpersonateMetric = Metric & { fields: ImpersonateFieldsMetric }
```

Example of queries and results:

```
SELECT
    payload.account,
    payload.kind,
    payload.fields.date,
    payload.fields.target.email as target_email,
    payload.fields.target.buy_org_id as target_buy_org,
    payload.fields.target.cost_center_id as target_cc

FROM "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
and payload.kind in ('impersonate-user-graphql-event')
order by ingestion_time desc
```

![image](https://github.com/vtex-apps/b2b-organizations-graphql/assets/1576737/a2da6bea-ee3a-4306-a2fd-bfe9e5b8e882)

```
SELECT
    payload.account,
    payload.kind,
    payload.fields.date,
    payload.fields.user.email as user_email,
    payload.fields.user.buy_org_id as user_buy_org,
    payload.fields.user.cost_center_id as user_cc,
    payload.fields.target.email as target_email,
    payload.fields.target.buy_org_id as target_buy_org,
    payload.fields.target.cost_center_id as target_cc

FROM "vtex"."schemaless"."b2b_suite_buyerorg_data_raw"
WHERE payload.name = 'b2b-suite-buyerorg-data'
and payload.kind in ('impersonate-b2b-user-graphql-event')
order by ingestion_time desc
```
![image](https://github.com/vtex-apps/b2b-organizations-graphql/assets/1576737/162e0feb-c4b5-4907-8631-c33012762ce9)


#### How does this PR make you feel?

![](https://media.giphy.com/media/PFgqgWiifz879nqUBO/giphy.gif)
